File: elixir/apps/arc_cli/lib/arc/cli/lists.ex
defmodule Arc.CLI.Lists do
  @moduledoc """
  Saved peer lists, per tool namespace, on this machine.

    arc lists add <tool> <name> <peer>...   add peers to a list
    arc lists rm <tool> <name> [<peer>...]  remove peers, or the list
    arc lists ls <tool> [<name>]            show lists, or one list

  A list lives at `~/.config/arc/lists/<tool>/<name>`, one peer per line.
  When a command of that tool resolves peers, a value that names a list
  expands to its members.
  """

  @default_dir Path.join(["~", ".config", "arc", "lists"])
  @name ~r/^[a-z0-9][a-z0-9._-]*$/

  @doc "Members of list `name` for `namespace`, or nil when there is no such list."
  @spec expand(String.t(), String.t()) :: [String.t()] | nil
  def expand(namespace, name) when is_binary(namespace) and is_binary(name) do
    if Regex.match?(@name, name) do
      case File.read(path(namespace, name)) do
        {:ok, text} -> String.split(text, "\n", trim: true)
        _ -> nil
      end
    end
  end

  def run(["add", namespace, name | peers]) when peers != [] do
    with :ok <- valid_name(name) do
      members = Enum.uniq((expand(namespace, name) || []) ++ peers)
      write(namespace, name, members)
      IO.puts("#{namespace}/#{name}: #{Enum.join(members, " ")}")
    end
  end

  def run(["rm", namespace, name]) do
    with :ok <- valid_name(name) do
      File.rm(path(namespace, name))
      IO.puts("removed #{namespace}/#{name}")
    end
  end

  def run(["rm", namespace, name | peers]) do
    with :ok <- valid_name(name) do
      members = (expand(namespace, name) || []) -- peers
      write(namespace, name, members)
      IO.puts("#{namespace}/#{name}: #{Enum.join(members, " ")}")
    end
  end

  def run(["ls", namespace]) do
    case File.ls(Path.expand(Path.join(dir(), namespace))) do
      {:ok, names} -> Enum.each(Enum.sort(names), &IO.puts("#{namespace}/#{&1}"))
      _ -> IO.puts("no lists for #{namespace}")
    end
  end

  def run(["ls", namespace, name]) do
    case expand(namespace, name) do
      nil -> error("no list #{namespace}/#{name}")
      members -> Enum.each(members, &IO.puts/1)
    end
  end

  def run(_) do
    IO.puts("""
    arc lists commands

      lists add <tool> <name> <peer>...
      lists rm <tool> <name> [<peer>...]
      lists ls <tool> [<name>]
    """)
  end

  defp write(namespace, name, members) do
    file = path(namespace, name)
    File.mkdir_p!(Path.dirname(file))
    File.write!(file, Enum.map_join(members, "", &(&1 <> "\n")))
  end

  defp valid_name(name) do
    if Regex.match?(@name, name), do: :ok, else: error("invalid list name #{name}")
  end

  defp dir, do: Application.get_env(:arc_cli, :lists_dir, @default_dir)
  defp path(namespace, name), do: Path.expand(Path.join([dir(), namespace, name]))

  @spec error(String.t()) :: no_return()
  defp error(msg) do
    IO.puts(:stderr, "error: #{msg}")
    Arc.CLI.Exit.halt(1)
  end
end

File: lists/lists.go
// Package lists holds the saved sets of peers of one machine.
//
// A list lives at ~/.config/arc/lists/<tool>/<name>, one peer for each line.
// A command of that tool that names a list sends to every peer in it.
package lists

import (
	"errors"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

// ErrInvalidName reports a name that a list may not take.
var ErrInvalidName = errors.New("lists: a name starts with a letter or a digit, and holds letters, digits, dot, dash and underscore")

var namePattern = regexp.MustCompile(`^[a-z0-9][a-z0-9._-]*$`)

// Store holds the lists of one machine.
type Store struct {
	// Dir is the directory of ARC, for example ~/.config/arc.
	Dir string
}

// DefaultStore returns the store of this user.
func DefaultStore() (*Store, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil, err
	}
	return &Store{Dir: filepath.Join(home, ".config", "arc")}, nil
}

func (s *Store) path(tool, name string) string {
	return filepath.Join(s.Dir, "lists", tool, name)
}

// Members returns the peers of one list, and nothing when there is no such
// list.
func (s *Store) Members(tool, name string) []string {
	if !namePattern.MatchString(tool) || !namePattern.MatchString(name) {
		return nil
	}

	data, err := os.ReadFile(s.path(tool, name))
	if err != nil {
		return nil
	}

	var members []string
	for _, line := range strings.Split(string(data), "\n") {
		if line = strings.TrimSpace(line); line != "" {
			members = append(members, line)
		}
	}
	return members
}

// Add puts peers in a list, and returns what the list holds now.
func (s *Store) Add(tool, name string, peers []string) ([]string, error) {
	if !namePattern.MatchString(tool) || !namePattern.MatchString(name) {
		return nil, ErrInvalidName
	}

	held := s.Members(tool, name)
	seen := map[string]bool{}
	for _, peer := range held {
		seen[peer] = true
	}

	for _, peer := range peers {
		if peer = strings.TrimSpace(peer); peer != "" && !seen[peer] {
			seen[peer] = true
			held = append(held, peer)
		}
	}
	return held, s.write(tool, name, held)
}

// Remove takes peers out of a list. With no peers, the list itself goes.
func (s *Store) Remove(tool, name string, peers []string) ([]string, error) {
	if !namePattern.MatchString(tool) || !namePattern.MatchString(name) {
		return nil, ErrInvalidName
	}

	if len(peers) == 0 {
		return nil, os.Remove(s.path(tool, name))
	}

	drop := map[string]bool{}
	for _, peer := range peers {
		drop[strings.TrimSpace(peer)] = true
	}

	var held []string
	for _, peer := range s.Members(tool, name) {
		if !drop[peer] {
			held = append(held, peer)
		}
	}
	return held, s.write(tool, name, held)
}

// Names returns the lists of one tool, in name order.
func (s *Store) Names(tool string) []string {
	if !namePattern.MatchString(tool) {
		return nil
	}

	files, err := os.ReadDir(filepath.Join(s.Dir, "lists", tool))
	if err != nil {
		return nil
	}

	var names []string
	for _, file := range files {
		if !file.IsDir() && namePattern.MatchString(file.Name()) {
			names = append(names, file.Name())
		}
	}
	sort.Strings(names)
	return names
}

func (s *Store) write(tool, name string, members []string) error {
	path := s.path(tool, name)
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}

	var body strings.Builder
	for _, member := range members {
		body.WriteString(member)
		body.WriteByte('\n')
	}
	return os.WriteFile(path, []byte(body.String()), 0o600)
}
